README.md
# pucks

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-08-05
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-08-05
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/pucks/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/pucks/compare/0.1.0...0.1.1


test/pucks/raytracer_test.clj
(ns pucks.raytracer-test
  (:require [clojure.test :refer :all]
            [pucks.parts :refer :all]
            [pucks.raytracer :refer :all]
            [clojure.core.matrix :as m]
            [same.core :refer [ish? zeroish?]]))


(deftest reflect-test
  (testing "vector reflection"
    (is (ish? [-1 0 0] (reflect [1 0 0] [-1 0 0])))))


(deftest mirrorsphere-normal
  (let [mirrorsphere0 (map->Sphere {:radius 5 :V [0 0 0]})]
    (testing "normals for X, Y, and Z"
      (is (ish? [1 0 0] (normal mirrorsphere0 [5 0 0])))
      (is (ish? [0 1 0] (normal mirrorsphere0 [0 5 0])))
      (is (ish? [0 0 1] (normal mirrorsphere0 [0 0 5]))))))



(deftest mirrorsphere-distance
  (testing "distance along x axis"
    (let [mirrorsphere0 (map->Sphere {:radius 5 :V [15 0 0]})
          ray0          (->Ray [0 0 0] [1 0 0])]
      (is (ish? 10 (distance mirrorsphere0 ray0)))))

  (testing "distance along x axis starting from 1"
    (let [mirrorsphere0 (map->Sphere {:radius 5 :V [15 0 0]})
          ray0          (->Ray [1 0 0] [1 0 0])]
      (is (ish? 9 (distance mirrorsphere0 ray0)))))

  (testing "infinite distance along x axis"
    (let [mirrorsphere0 (map->Sphere {:radius 5 :V [0 15 0]})
          ray0          (->Ray [0 0 0] [-1 0 0])]
      (is (ish? ##Inf (distance mirrorsphere0 ray0)))))
  
  (testing "distance along -x axis"
    (let [mirrorsphere0 (map->Sphere {:radius 5 :V [-15 0 0]})
          ray0          (->Ray [0 0 0] [-1 0 0])]
      (is (ish? 10 (distance mirrorsphere0 ray0)))))

  (testing "infinite distance along -x axis"
    (let [mirrorsphere0 (map->Sphere {:radius 5 :V [15 0 0]})
          ray0          (->Ray [0 0 0] [-1 0 0])]
      (is (ish? ##Inf (distance mirrorsphere0 ray0)))))
  
  (testing "zero distance is infinity"
    (let [mirrorsphere0 (map->Sphere {:radius 5 :V [15 0 0]})
          ray0          (->Ray [10 0 0] [-1 0 0])]
      (is (ish? ##Inf (distance mirrorsphere0 ray0))))))





(deftest mirror-normal
  (testing "normal in positive z direction"
    (let [mirror0 (->Mirror [0 0 0] [1 0 0] [1 1 0] [0 1 0])]
      (is (ish? [0 0 1] (normal mirror0))))))

(deftest mirror-distance
  (testing "distance test in +x direction"
    (let [mirror0 (->Mirror [10 0 0] [10 1 0] [10 1 1] [10 0 1])
          ray0    (->Ray [0 0 0] [1 0 0])]
      (is (ish? 10 (distance mirror0 ray0)))))
  (testing "infinite distance test in +x direction"
    (let [mirror0 (->Mirror [10 1 1] [10 2 1] [10 2 2] [10 1 2])
          ray0    (->Ray [0 0 0] [1 0 0])]
      (is (ish? ##Inf (distance mirror0 ray0))))))


src/pucks/parts.clj
(ns pucks.parts
  (:require [pucks.colors :refer :all]
            [scad-clj.model :refer :all]
            [pucks.scad-extended :refer :all]
            [clojure.core.matrix :as m]
            [clojure.math :refer [acos atan2 sqrt pow cos sin]]
            [pucks.raytracer :refer :all]))


;; Protocols

(defprotocol MechanicalPart
  (build [this]))


(defprotocol Transform
  (affinetransform [this M])
  (scale- [this [sx sy sz]])
  (translate- [this [dx dy dz]])
  (rotatex [this theta])
  (rotatey [this theta])
  (rotatez [this theta])
  (rotate- [this [rx ry rz]]))




;; All Parts
(defrecord Sphere [radius V])

(defrecord Prism [normal])

(defrecord Mirror [V0 V1 V2 V3])






;; Translations

(extend-type Sphere
  Transform
  (translate- [this V1]
    (->Sphere (:radius this) (m/add (:V this) V1))))


(extend-type Mirror
  Transform
  (affinetransform [this M]
    ;; affine transform   
    ;; |x'|      |1 0 0 tx||x|  
    ;; |y'|   =  |0 1 0 ty||y|
    ;; |z'|      |0 0 1 tz||z|
    ;; |1 |      |0 0 0 1 ||1|
    (apply ->Mirror
           (vec (for [[x y z] (vals this)
                      :let [[x' y' z' _] (m/mmul M [x y z 1])]]
                  [x' y' z']))))
  (scale- [this [sx sy sz]]
    (affinetransform this [[sx 0  0  0]
                           [0  sy 0  0]
                           [0  0  sz 0]
                           [0  0  0  1]]))
  (translate- [this [dx dy dz]]
    (affinetransform this [[1 0 0 dx]
                           [0 1 0 dy]
                           [0 0 1 dz]
                           [0 0 0 1 ]]))
  (rotatex [this theta]
    (affinetransform this [[1  0              0           0]
                           [0 (cos theta)     (- (sin theta)) 0]
                           [0 (sin theta) (cos theta) 0]
                           [0  0              0           1]]))
  (rotatey [this theta]
    (affinetransform this [[(cos theta) 0  (sin theta) 0]
                           [0           1  0               0]
                           [(- (sin theta)) 0  (cos theta)     0]
                           [0           0  0               1]]))
  (rotatez [this theta]
    (affinetransform this [[(cos theta) (- (sin theta)) 0          0]
                           [(sin theta) (cos theta)     0          0]
                           [0           0               1          0]
                           [0           0               0          1]]))
  (rotate- [this [rx ry rz]]
    (-> this (rotatex rx) (rotatey ry) (rotatez rz))))







;; Constructors

(defn make-mirror
  ([length width] (make-mirror length width {:center true}))
  ([length width & [{:keys [center]} options]]
   (let [max-y (/ length 2) max-z (/ width 2)
         m0 (->Mirror [0 -0.5 -0.5] [0 -0.5 0.5] [0 0.5 0.5] [0 0.5 -0.5])]
     (if (true? center) (scale- m0 [0 length width])
         (translate- (make-mirror length width {:center true})
                     [0 (/ length 2) (/ width 2) ])))))




;; Mechanical parts


(defrecord Galvo
    [base-radius base-height
     trunk-radius trunk-height
     spindle-radius spindle-height
     mirror-width mirror-height mirror-thickness mirror-translation-z
     pcb-width pcb-length pcb-thickness pcb-translation-z
     total-height]
  MechanicalPart
  (build [this]
    (let [base    (->> (cylinder base-radius base-height :center false)
                       (with-fn 30)
                       (color black))
          trunk   (->> (cylinder trunk-radius trunk-height :center false)
                       (with-fn 30)
                       (color black))
          spindle (->> (cylinder spindle-radius spindle-height :center false)
                       (with-fn 20)
                       (color lightgray))
          mirror  (->> (cube mirror-width mirror-thickness mirror-height)
                       (color silver))
          pcb     (->> (cube pcb-length pcb-width pcb-thickness :center false)
                       (translate [0 (/ pcb-width -2) 0])
                       (color wheat))]
      (union base trunk spindle
             (->> pcb (translate [0 0 pcb-translation-z]))
             (->> mirror (translate [0 0 mirror-translation-z]))))))


(defrecord GalvoHolder [L-height L-width L-thickness height galvo-x galvo-y]
  MechanicalPart
  (build [this]
    (let [slot (->> (cube 2 height L-thickness :center false)
                    (rotate [half half 0])
                    (translate [0 0 25]))
          base (extrude-linear {:height height :center false :convexity 10}
                               (polygon [[0,0] [L-width 0] [L-width L-thickness]
                                         [L-thickness L-thickness]
                                         [L-thickness L-height] [0 L-height]]))
          move-x (fn [obj] (->> obj
                                (rotate [-quarter 0 0])
                                (translate [40 -25 30])))
          move-y (fn [obj] (->> obj
                                (rotate [0 quarter 0])
                                (rotate [-quarter 0 0])
                                (translate [-25 40 30])))]
      (union (difference (->> base (color lightslategray))
                         (->> galvo-x (scale [1.1 1.1 1.1]) move-x)
                         (->> slot move-x)
                         (->> galvo-y (scale [1.1 1.1 1.1]) move-y)
                         (->> slot move-y))
             (->> galvo-x move-x)
             (->> galvo-y move-y)))))


(defrecord OpticalTable [length width thickness hole-radius spacing]
  MechanicalPart
  (build [this]
    (color silver
           (difference
            (cube length width thickness :center false)
            (for [x (range spacing length spacing)
                  y (range spacing width spacing)]
              (->> (cylinder hole-radius (inc thickness) :center false)
                   (with-fn 12)
                   (translate [x y 0])))))))

(defrecord LaserDiode [base-radius base-height
                       head-radius head-height]
  MechanicalPart
  (build [this]
    (let [base  (cylinder base-radius base-height :center false)
          head  (cylinder head-radius head-height :center false)
          glass (cylinder 2 2 :center false) 

          prong (->> (cylinder 0.8 8 :center false)
                     (with-fn 10)
                     (mirror [0 0 1])
                     (translate [0 4 0]))]
      (union
       base
       (->> head (translate [0 0 base-height]) (color silver))
       (->> glass (translate [0 0 (+ base-height 6.1)]) (color blue))
       (->> prong (rotate [0 0 (* tau 1/3)]))
       (->> prong (rotate [0 0 (* tau 2/3)]))
       (->> prong (rotate [0 0 (* tau 3/3)]))))))



(defrecord LensParabola [a b]
  MechanicalPart
  (build [this]
    (let [parabola (fn [x] (+ (* a x x) (* b x)))
          xs       (range 0 11)
          ys       (map parabola xs)
          lasty    (last ys)]
      (->> (polygon (conj (m/transpose [xs ys]) [0 lasty]))
           (translate [0 (- lasty) 0])
           (extrude-rotate)
           (rotate [0 -quarter 0])
           (color (conj silver 0.3))))))



(defrecord LaserBeam [points radius beamcolor]
  MechanicalPart
  (build [_]
    (->> (with-fn 6 (circle (or radius 0.5)))
         (extrude-points points)
         (color (or beamcolor (conj blue 0.7))))))







(extend-type Mirror
  MechanicalPart
  (build [this]
    (let [{:keys [V0 V1 V2 V3]} this
          anti-norm       (m/mul -1 (m/normalise (normal this)))
          [V4 V5 V6 V7]   (map (partial m/add anti-norm) [V0 V1 V2 V3])
          mirror          (polyhedron [V0 V1 V2 V3 V4 V5 V6 V7]
                                      [[0 1 2 3] ;; bottom
                                       [4 5 1 0] ;; front
                                       [7 6 5 4] ;; top
                                       [5 6 2 1] ;; right
                                       [6 7 3 2] ;; back
                                       [7 4 0 3] ;; left
                                       ])]
      (union (->> mirror (color (conj silver 0.9)))
             (->> mirror (translate anti-norm) (color black))))))



(extend-type Prism
  MechanicalPart
  (build [this]
    (let [points [[0 0] [0 10] [10 0]]
          normal  [1 1 0]]
      (->> (polygon points)
           (extrude-linear {:height 10})
           (color (conj silver 0.3))))))


(extend-type Sphere
  MechanicalPart
  (build [this]
    (->> (with-fn 32 (sphere (:radius this)))
         (translate (:V this))
         (color (conj silver 0.9)))))



















(comment "FOR ALGEBRAIC CONSTRAINTS"

;; Phrozen sonic 4k:  120 x 68 x 130 mm

(let [m {:base-radius 20 :base-height 20
         :trunk-radius 10 :trunk-height 20
         :spindle-radius 5 :spindle-height 5
         :mirror-width 20 :mirror-height 20 :mirror-thickness 0.5
         :pcb-height 10 :pbc-width 20 :pcb-thickness 0.5
         :total-height 61}]
  (logic/run* [q]
    (logic/== (:pcb-width q) (:base-radius m))
    ;(logic/== {:hello 3 :goodbye 2 :nest 0} q)
    ))

[:base-radius :base-height
 :trunk-radius :trunk-height
 :spindle-radius :spindle-height
 :mirror-width :mirror-height :mirror-thickness
 :pcb-height :pbc-width :pcb-thickness
 :total-height]

;; base-height + trunk-height + spindle-height + mirror-height = total-height
;; pcb-width = base-radius


;; Probably not going to use 
(defmacro defpart [] nil)
 )


src/pucks/colors.clj
(ns pucks.colors)

(def -colorsmap
  {:aquamarine [127 255 212 1],
   :lime [0 255 0 1],
   :deepskyblue [0 191 255 1],
   :darksalmon [233 150 122 1],
   :antiquewhite [250 235 215 1],
   :mediumturquoise [72 209 204 1],
   :slategrey [112 128 144 1],
   :slategray [112 128 144 1],
   :sienna [160 82 45 1],
   :orange [255 165 0 1],
   :navajowhite [255 222 173 1],
   :lavenderblush [255 240 245 1],
   :firebrick [178 34 34 1],
   :orangered [255 69 0 1],
   :transparent [0 0 0 0],
   :palevioletred [219 112 147 1],
   :lawngreen [124 252 0 1],
   :seashell [255 245 238 1],
   :lightpink [255 182 193 1],
   :darkolivegreen [85 107 47 1],
   :aliceblue [240 248 255 1],
   :gray [128 128 128 1],
   :lightsteelblue [176 196 222 1],
   :whitesmoke [245 245 245 1],
   :darkgoldenrod [184 134 11 1],
   :tan [210 180 140 1],
   :bisque [255 228 196 1],
   :white [255 255 255 1],
   :lightgreen [144 238 144 1],
   :darkseagreen [143 188 143 1],
   :crimson [220 20 60 1],
   :darkslategray [47 79 79 1],
   :mistyrose [255 228 225 1],
   :chocolate [210 105 30 1],
   :yellow [255 255 0 1],
   :cadetblue [95 158 160 1],
   :navy [0 0 128 1],
   :ghostwhite [248 248 255 1],
   :dimgrey [105 105 105 1],
   :rebeccapurple [102 51 153 1],
   :seagreen [46 139 87 1],
   :green [0 128 0 1],
   :mediumseagreen [60 179 113 1],
   :indigo [75 0 130 1],
   :olivedrab [107 142 35 1],
   :cyan [0 255 255 1],
   :peachpuff [255 218 185 1],
   :limegreen [50 205 50 1],
   :mediumslateblue [123 104 238 1],
   :violet [238 130 238 1],
   :sandybrown [244 164 96 1],
   :yellowgreen [154 205 50 1],
   :mediumspringgreen [0 250 154 1],
   :steelblue [70 130 180 1],
   :rosybrown [188 143 143 1],
   :cornflowerblue [100 149 237 1],
   :ivory [255 255 240 1],
   :lightgoldenrodyellow [250 250 210 1],
   :salmon [250 128 114 1],
   :darkcyan [0 139 139 1],
   :peru [205 133 63 1],
   :cornsilk [255 248 220 1],
   :lightslategray [119 136 153 1],
   :blueviolet [138 43 226 1],
   :forestgreen [34 139 34 1],
   :lightseagreen [32 178 170 1],
   :gold [255 215 0 1],
   :gainsboro [220 220 220 1],
   :darkorchid [153 50 204 1],
   :burlywood [222 184 135 1],
   :lightskyblue [135 206 250 1],
   :chartreuse [127 255 0 1],
   :snow [255 250 250 1],
   :moccasin [255 228 181 1],
   :honeydew [240 255 240 1],
   :aqua [0 255 255 1],
   :darkred [139 0 0 1],
   :mediumorchid [186 85 211 1],
   :lightsalmon [255 160 122 1],
   :saddlebrown [139 69 19 1],
   :wheat [245 222 179 1],
   :springgreen [0 255 127 1],
   :lightslategrey [119 136 153 1],
   :darkblue [0 0 139 1],
   :powderblue [176 224 230 1],
   :turquoise [64 224 208 1],
   :blanchedalmond [255 235 205 1],
   :papayawhip [255 239 213 1],
   :slateblue [106 90 205 1],
   :lightblue [173 216 230 1],
   :skyblue [135 206 235 1],
   :red [255 0 0 1],
   :lightyellow [255 255 224 1],
   :blue [0 0 255 1],
   :palegreen [152 251 152 1],
   :greenyellow [173 255 47 1],
   :khaki [240 230 140 1],
   :maroon [128 0 0 1],
   :darkgrey [169 169 169 1],
   :midnightblue [25 25 112 1],
   :floralwhite [255 250 240 1],
   :deeppink [255 20 147 1],
   :paleturquoise [175 238 238 1],
   :darkkhaki [189 183 107 1],
   :azure [240 255 255 1],
   :indianred [205 92 92 1],
   :darkviolet [148 0 211 1],
   :mediumpurple [147 112 219 1],
   :fuchsia [255 0 255 1],
   :coral [255 127 80 1],
   :mediumvioletred [199 21 133 1],
   :lemonchiffon [255 250 205 1],
   :mediumblue [0 0 205 1],
   :darkmagenta [139 0 139 1],
   :goldenrod [218 165 32 1],
   :darkorange [255 140 0 1],
   :orchid [218 112 214 1],
   :plum [221 160 221 1],
   :pink [255 192 203 1],
   :teal [0 128 128 1],
   :magenta [255 0 255 1],
   :lightgrey [211 211 211 1],
   :purple [128 0 128 1],
   :dodgerblue [30 144 255 1],
   :darkturquoise [0 206 209 1],
   :mintcream [245 255 250 1],
   :hotpink [255 105 180 1],
   :thistle [216 191 216 1],
   :royalblue [65 105 225 1],
   :darkgreen [0 100 0 1],
   :darkslateblue [72 61 139 1],
   :silver [192 192 192 1],
   :darkgray [169 169 169 1],
   :grey [128 128 128 1],
   :oldlace [253 245 230 1],
   :mediumaquamarine [102 205 170 1],
   :brown [165 42 42 1],
   :lightgray [211 211 211 1],
   :olive [128 128 0 1],
   :lightcoral [240 128 128 1],
   :tomato [255 99 71 1],
   :lightcyan [224 255 255 1],
   :linen [250 240 230 1],
   :darkslategrey [47 79 79 1],
   :lavender [230 230 250 1],
   :dimgray [105 105 105 1],
   :palegoldenrod [238 232 170 1],
   :beige [245 245 220 1],
   :black [51 51 51 1],
   :deepblack [0 0 0 1]})



(defmacro -assign-colors []
  `(do ~@(map (fn [[k [r g b a]]]
                `(def ~(symbol k) [~(/ r 255) ~(/ g 255) ~(/ b 255)]))
              -colorsmap)))

(comment (macroexpand-1 '(-assign-colors)))
(-assign-colors)


src/pucks/raytracer.clj
(ns pucks.raytracer
  (:use  [scad-clj.scad]
         [scad-clj.model :rename {import scadimport, use scaduse}]
         [numeric.expresso.core])
  (:require
   [pucks.colors :refer :all]
   [clojure.core.matrix :as m]
   [pucks.scad-extended :refer :all]
   [clojure.math :refer [acos atan2 sqrt pow]]
   [same.core :refer [zeroish? not-zeroish?]]))


;; Ray tracing primer
;; https://www.cl.cam.ac.uk/teaching/1999/AGraphHCI/SMAG/node2.html

;; Conventions
;; E - origin point
;; D - offset vector
;; P(t) = E + tD, where t>=0

;; V0, V1, V2, V3... vectors are always zero indexed




;;;;;;;;;;;;;;;;;;;;;;;
;; Protocols

(defprotocol OpticalPart
  (distance [this ray]
    "calculate the distance between [this-obj] and a ray")
  (point-on? [this point]
    "test if this point lies on the surface of [this-obj]")
  (normal [this] [this point]
    "calculate the normal vector of [this-obj] at a point"))


(defprotocol RayTrace
  (trace [this elements]
    "returns a lazy seq of all points along a reflective path")
  (get-closest [this elements]
    "returns the closest element, and its distance")
  (rayproject [this distance]
    "project a ray by a certain distance, return two points"))




;;;;;;;;;;;;;;;;;;;;;;;
;; Math

(defn reflect
  "vector reflection. r = d - 2(d . n) n
  d = incoming normal; n = surface normal"
  [d n]
  (m/sub d (m/mul 2 (m/dot d n) n)))


(defn solve-quad
  "solves roots for quadratics
  determinant = b^2 - 4ac
  + = 2 solutions, 0 = one solution, - = complex solutions
  "
  [a b c]
  (-> "a*x**2 + b*x + c = 0"
      parse-expression
      (substitute {'a a 'b b 'c c})
      (as-> expr (solve 'x expr))))

(def solve-fast-quad
  (->> "a*x**2 + b*x + c = 0"
       parse-expression
       (solve 'x)
       first ;; potentially problematic
       (compile-expr [a b c])))





;;;;;;;;;;;;;;;;;;;;;;;
;; Raytracing


(defrecord Ray [rayorigin raynormal]
  RayTrace
  (get-closest [this elements]
    (->> elements
         (map (fn [element] [element (distance element this)]))
         (sort-by second) first))
  (rayproject [this distance]
    [rayorigin (m/add rayorigin (m/dot raynormal distance))])
  (trace [this elements]
    (lazy-seq
     (cons rayorigin
           (let [[closest distance-to] (get-closest this elements)]
             (if (infinite? distance-to) [(second (rayproject this 400))]
                 (let [[V0 V1] (rayproject this distance-to)]
                   (trace (->Ray V1 (reflect raynormal (normal closest V1)))
                          elements))))))))







;;;;;;;;;;;;;;;;;;;;;;;
;; Optical Parts


;; sphere equation
;; (x - Xc)^2 + (y-Yc)^2 + (z-Zc)^2 - R^2 = 0
(extend-type pucks.parts.Sphere
  OpticalPart
  (distance [this ray]
    (let [sphere-to-ray (m/sub (:rayorigin ray) (:V this))
          a 1
          b (* 2 (m/dot (:raynormal ray) sphere-to-ray))
          c (- (m/dot sphere-to-ray sphere-to-ray) (pow (:radius this) 2))
          discriminant (- (pow b 2) (* 4 a c))
          distance (/ (- (- b) (sqrt discriminant)) (* 2 a)) 
          distance2 (solve-fast-quad a b c)]
      (cond (NaN? distance)  ##Inf
            (neg? distance)  ##Inf
            (zero? distance) ##Inf
            :else            distance)))
  (point-on? [this point] "not implemented")
  (normal [this point] (m/normalise (m/sub point (:V this)))))



(extend-type pucks.parts.Mirror
  OpticalPart
  (distance [this ray]
    (let [{:keys [V0 V1 V2 V3]} this
          Q V0, E (:rayorigin ray), D (:raynormal ray), N (normal this)
          t (m/div (m/dot N (m/sub Q E)) (m/dot N D))
          P0 (m/add (m/dot t D) E)]
      (if (and (not-zeroish? t) (point-on? this P0)) t ##Inf)))
  (point-on? [this P0]
    (let [{:keys [V0 V1 V2 V3]} this]
      (let [U (m/sub V1 V0) W (m/sub V3 V0)]
        (and (<= (m/dot U V0) (m/dot U P0) (m/dot U V1))
             (<= (m/dot W V0) (m/dot W P0) (m/dot W V3))))))
  (normal
    ([this]
     (let [{:keys [V0 V1 V2]} this]
       (m/normalise (m/cross (m/sub V2 V1) (m/sub V0 V1)))))
    ([this point] (normal this))))






src/pucks/scad_extended.clj
(ns pucks.scad-extended
  (:use  [scad-clj.scad]
         [scad-clj.model :rename {import scadimport, use scaduse}])
  (:require
   [clojure.core.matrix :as m]
   [clojure.math.combinatorics :as combo]
   [clojure.math :refer [acos atan2 sqrt pow]]

   ;; cad
   [pucks.colors :refer :all]))


(def quarter (deg->rad 90))
(def -quarter (- (deg->rad 90)))
(def half (deg->rad 180))



(defn rotate-vector
  "Rotate a part from [0 0 1] to align perfectly with a vector"
  [[x y z] part]
  (let [magnitude (m/length [x y z])]
    (rotate [0 (acos (/ z magnitude)) (atan2 y x)] part)))



(defn extrude-vector
  "extrudes a 2d shape into a vector"
  ([[x y z]] (extrude-vector [x y z] (circle 0.5)))
  ([[x y z] shape]
   (when-not (m/zero-matrix? [x y z])
     (let [magnitude (m/length [x y z])]
       (->> shape
            (extrude-linear {:height magnitude :center false})
            (rotate-vector [x y z]))))))



(def extrude-vector-example
  (map #(color (m/div (m/add % 10) 20) (extrude-vector %))
       (combo/permuted-combinations
        [-10 -10 -10 0 0 0 10 10 10] 3)))



(defn extrude-points
  "extrudes a shape through a collection of points in 3d"
  ([points] (extrude-points points (circle 0.5)))
  ([points startshape]
   (let [pairs (partition 2 1 points)]
     (for [[V0 V1] pairs
           :let [V' (m/sub V1 V0)]]
       (->> (extrude-vector V' startshape) (translate V0))))))


(defn extrude-along-path
  "extrudes a shape through a collection of points in 3d, with no sharp edges
  this function is not done.
  "
  ([points] (extrude-along-path points (circle 0.5)))
  ([points startshape]
   (union
    (let [pairs   (partition 2 1 (vec points))
                hulls  (for [[V0 V1] pairs ;; hull will only work on 3d shapes
                             :let [V' (m/sub V1 V0)
                                   N  (m/normalise V')]]
                         (->> (with-fn 30 (circle 0.4))
                              (extrude-linear {:height 0.1 :center false})
                              (rotate-vector N)
                              (translate V0)))]
      (map #(apply hull %) (partition 2 1 hulls))))))


(def extrude-along-path-example
  (let [f      (fn [x] (+ (* -2 (pow x 2)) 6))
        points (for [x (range 0 5 1)
                     :let [y (f x)]]
                 [x y 0])]
    (extrude-along-path points)))




src/pucks/core.clj
(ns pucks.core
  (:use  [scad-clj.scad]
         [scad-clj.model :rename {import scadimport, use scaduse}])
  (:require
   ;; math
   [clojure.core.logic :as logic]
   [clojure.core.matrix :as m]
   [clojure.math.combinatorics :as combo]
   [clojure.math :refer [acos atan2 sqrt pow]]

   ;; traversal
   [clojure.walk :refer [postwalk]]

   ;; cad
   [pucks.colors :refer :all]
   [pucks.scad-extended :refer :all]
   [pucks.parts :refer :all]
   [pucks.raytracer :refer :all :reload :all]))


;;UNITS mm




;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Parts

(def build-plate (square 120 68 :center false))
(def build-volume (->> (cube 120 68 130 :center false)
                       (color [1 0 0 0.2])))

(def puck (cube 3 20 20 :center false))

(def table (build (map->OpticalTable {:length 500 :width 250 :thickness 10
                                      :hole-radius 4 :spacing 50})))


(def galvo-1 (build (map->Galvo
                     {:base-radius 20 :base-height 20
                      :trunk-radius 10 :trunk-height 50
                      :spindle-radius 2 :spindle-height 55
                      :mirror-width 10 :mirror-height 20 :mirror-thickness 0.5
                      :pcb-width 40 :pcb-length 50 :pcb-thickness 1
                      :mirror-translation-z 63 :pcb-translation-z 15
                      :total-height 61})))

(def galvoholder (build (map->GalvoHolder
                         {:L-height 80 :L-width 80 :L-thickness 15
                          :height 70 :galvo-x galvo-1 :galvo-y galvo-1})))

(def bluelaserdiode (build (map->LaserDiode
                        {:base-radius 10 :base-height 3
                         :head-radius 6 :head-height 8})))

(def redlaserdiode (build (map->LaserDiode
                        {:base-radius 10 :base-height 3
                         :head-radius 6 :head-height 8})))


(def lensparabola (build (map->LensParabola
                          {:a 1/60 :b 1/20 :focal-length 20})))

;; (def prism (build (map->Prism {})))

(def bluebeam
  (map->LaserBeam {:radius 0.5 :points [[0 0 0] [0 0 200]]
                   :beamcolor (conj blue 0.7)}))

(def redbeam
  (map->LaserBeam {:radius 0.5 :points [[0 0 0] [0 0 200]]
                   :beamcolor (conj red 0.7)}))


(def optical-mirror (map->Mirror {:length 20 :width 10
                                  :thickness 0.2 :normal [1 1 1]}))






;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Assembly


(def MainAssembly
  (union
   ;; hardware
   (->> galvoholder (rotate [0 0 half]) (translate [400 150 0]))
   (->> table (translate [0 0 -10]))

   ;; lasers
   (->> bluelaserdiode (rotate [0 quarter 0]) (translate [100 100 20]))
   (->> redlaserdiode (rotate [0 quarter -quarter]) (translate [155 200 20]))


   (->> bluebeam build (rotate [0 quarter 0]) (translate [100 100 20]))
   (->> redbeam build (rotate [0 quarter -quarter]) (translate [155 200 20]))


   ;; optics
   (->> lensparabola (rotate [0 0 0]) (translate [130 100 20]))
   ;; (->> prism (rotate [0 0 0]) (translate [150 100 20]))
   )

  )



(defrecord BeamTestAssembly []
  MechanicalPart
  (build [_]
    (union
     ;; lasers
     (->> bluelaserdiode (rotate [0 quarter 0]))
     (->> redlaserdiode (rotate [0 quarter -quarter]))

     (->> bluebeam (rotate [0 quarter 0]))
     (->> redbeam (rotate [0 quarter -quarter]))

     ;; optics
     (build optical-mirror)
     ;; (->> lensparabola (rotate [0 0 0]) (translate [130 100 20]))
     ;; (->> prism (rotate [0 0 0]) (translate [150 100 20]))
     )))



(def raytraceexample
  (let [ray0 (->Ray [0 0 0] (m/normalise [1 0 0]))
        
        mirror0 (map->Sphere {:radius 20 :V [25 13 10]})
        mirror1 (map->Sphere {:radius 10 :V [27 -13 -10]})
        mirror2 (map->Sphere {:radius 20 :V [-50 -20 -5]})
        mirror3 (map->Sphere {:radius 15 :V [-50 -50 -65]})

        mirrorp0 (-> (make-mirror 15 5)
                     (rotate- [0 0.5 1.6])
                     (translate- [20 30 -22]))

        points (trace ray0 [mirror0 mirror1 mirror2 mirror3 mirrorp0])
        
        laserbeam (map->LaserBeam {:points points})]
    (union
     (map build [laserbeam mirror0 mirror1 mirror2 mirror3 mirrorp0]))))






;; Compile
(comment
  (spit "pucks.scad"
        (write-scad build-plate
                    (-% build-volume) puck)))

(spit "pucks.scad"
      (write-scad
       
       ;; extrude-along-path-example

       ;; raytraceexample
        MainAssembly
       
       ;; (build (->MainAssembly))
       ;;(build (->BeamTestAssembly))
       ))





